plugins/value-container-assignment/testData/diagnostics/localVariablesTest.kt

```diff
@@ -0,0 +1,51 @@
+// !DIAGNOSTICS: -UNUSED_PARAMETER,-UNUSED_VARIABLE, -ASSIGNED_BUT_NEVER_ACCESSED_VARIABLE
+// !RENDER_DIAGNOSTICS_FULL_TEXT
+
+annotation class ValueContainer
+
+@ValueContainer
+data class StringProperty(var v: String) {
+    fun assign(v: String) {
+        this.v = v
+    }
+    fun assign(v: StringProperty) {
+        this.v = v.get()
+    }
+    fun get(): String = v
+}
+
+fun test() {
+    run {
+        // Should not work with local val for different type
+        val property = StringProperty("OK")
+        <!VAL_REASSIGNMENT!>property<!> = <!TYPE_MISMATCH!>"Fail"<!>
```

Comment:
We had a discussion about that inside the team, due to a community comment from operator KEEP about `var`/`val` ambiguity. The comment was, that it might be error-prone since:
```
val property = StringProperty("OK")
property = StringProperty("OK") // this calls assign()

var property = StringProperty("OK")
property = StringProperty("OK") // this changes a reference
```

We then checked our use case, and it works for us if we forbid using it for local variables. And if we figure it would be good to support local variables, we can remove that limitation without a breaking change. The other way around is much harder.

What do you think about that?

Reply:
IMO there is no much difference between local, member and top-level properties, especially for scripts. Declaring such property as `var` can be error-prone in any place, not in local scopes. So I would add an additional checker which reports a warning if someone tries to declare `var` of type which contains your annotation